### PR TITLE
UCP: Migrate scalar function `Lower` from TiDB (#6940)

### LINKED ISSUE #6940: UCP: Migrate scalar function `Lower` from TiDB
## Description

Port the scalar function `Lower` from TiDB to coprocessor.

## Score 

* 50

## Mentor(s)

* @sticnarf 

## Recommended Skills

* Rust programming 

## Learning Materials

Already implemented expressions ported from TiDB

- https://github.com/tikv/tikv/tree/master/components/tidb_query/src/rpn_expr) 
- https://github.com/tikv/tikv/tree/master/components/tidb_query/src/expr)

### PR DESCRIPTION
**Pull request by Izumi39**
Signed-off-by: Jiannuan <gan.jiannuan@gmail.com>

### What problem does this PR solve?

Issue Number: close #6940 

Problem Summary: Migrate scalar function `Lower` from TiDB

### What is changed and how it works?

What's Changed:
Implemented scalar function `lower`

### Check List 

Tests 

- Unit test

### CODE REVIEW

**Comment by bodhif4** [comment]
Help me. CI Integration MySQL Test and Integration MySQL Test Cached failure.
### log
> [2020-05-10T14:30:56.616Z] time="2020-05-10T22:30:56+08:00" level=fatal msg="run test [gcol_blocked_sql_funcs] err: sql:create table t1 (a varchar(32), b bool generated always as (match a against ('sample text')) virtual);: run \"create table t1 (a varchar(32), b bool generated always as (match a against ('sample text')) virtual);\" around line 80 err, we need(299):\ncreate table t1 (a varchar(32), b bool generated always as (match a against ('sample text')) virtual);\n[parser:1064]You have an error in your SQL syntax; check the manual that corresponds to your TiDB version for the right syntax to use line 1 column 65 near \"match a against ('sample text')) virtua\nbut got(299):\ncreate table t1 (a varchar(32), b bool generated always as (match a against ('sample text')) virtual);\n[parser:1064]You have an error in your SQL syntax; check the manual that corresponds to your TiDB version for the right syntax to use line 1 column 67 near \"a against ('sample text')) virtual);\" \n\n"
[2020-05-10T14:30:56.616Z] tidb-server(PID: 39146) stopped
script returned exit code 1

> [2020-05-10T14:30:57.417Z] + set +e
> [2020-05-10T14:30:57.417Z] + killall -9 -r tidb-server

> [2020-05-10T14:30:57.417Z] tidb-server: no process found

> [2020-05-10T14:30:57.417Z] + killall -9 -r tikv-server

> [2020-05-10T14:30:57.417Z] + killall -9 -r pd-server

> [2020-05-10T14:30:57.417Z] + set -e

> script returned exit code 1

**Comment by Fullstop000** [comment]
@bodhif4 integration test is now broken on the master branch. You can just ignore it.

File: components/tidb_query_vec_expr/src/impl_string.rs [hunk 1]
```diff
@@ -1919,4 +1925,37 @@ mod tests {
...
+                Some("قاعدة البيانات".as_bytes().to_vec()),
+                Some("قاعدة البيانات".as_bytes().to_vec()),
+            ),
+            (None, None),
+        ];
+
+        for (arg, exp) in cases {
+            let output = RpnFnScalarEvaluator::new()
+                .push_param(arg.clone())
+                .evaluate(ScalarFuncSig::Upper)
```
Comment (Fullstop000) [inline components/tidb_query_vec_expr/src/impl_string.rs:1956]: Why use `Upper`?

Comment (bodhif4) [reply components/tidb_query_vec_expr/src/impl_string.rs]: Sorry. I forgot to modify the test code. Now fixed.

Comment (bodhif4) [reply components/tidb_query_vec_expr/src/impl_string.rs]: > Why use `Upper`?
Sorry. I forgot to modify the test code. Now fixed.

**Comment by bodhif4** [comment]
> @bodhif4 integration test is now broken on the master branch. You can just ignore it.

thanks

File: components/tidb_query_vec_expr/src/impl_string.rs [hunk 2]
```diff
@@ -492,6 +492,12 @@ pub fn to_base64(bs: &Option<Bytes>) -> Result<Option<Bytes>> {
     }
 }
 
+#[rpn_fn]
+#[inline]
+pub fn lower(arg: &Option<Bytes>) -> Result<Option<Bytes>> {
+    Ok(arg.as_ref().map(|b| b.to_vec()))
```
Comment (Fullstop000) [inline components/tidb_query_vec_expr/src/impl_string.rs:498]: The `lower` func in the normal version hasn't been separated into `lower` and `lower_utf8` yet. Would you like to post a PR to implement it :)? . You can refer to `upper` and `upper_utf8` in this PR #6280.

**Comment by bodhif4** [comment]
OK, thanks. I want to try it.

获取 Outlook for iOS<https://aka.ms/o0ukef>
________________________________
发件人: Fullstop000 <notifications@github.com>
发送时间: Monday, May 11, 2020 10:40:41 PM
收件人: tikv/tikv <tikv@noreply.github.com>
抄送: jiannuan <gan.jiannuan@gmail.com>; Mention <mention@noreply.github.com>
主题: Re: [tikv/tikv] UCP: Migrate scalar function `Lower` from TiDB (#6940) (#7780)


@Fullstop000 commented on this pull request.

________________________________

In components/tidb_query_vec_expr/src/impl_string.rs<https://github.com/tikv/tikv/pull/7780#discussion_r423089395>:

> @@ -492,6 +492,12 @@ pub fn to_base64(bs: &Option<Bytes>) -> Result<Option<Bytes>> {
     }
 }

+#[rpn_fn]
+#[inline]
+pub fn lower(arg: &Option<Bytes>) -> Result<Option<Bytes>> {
+    Ok(arg.as_ref().map(|b| b.to_vec()))


The lower func in the normal version hasn't been separated into lower and lower_utf8 yet. Would you like to post a PR to implement it :)? . You can refer to upper and upper_utf8 in this PR #6280<https://github.com/tikv/tikv/pull/6280>.

―

File: components/tidb_query_vec_expr/src/impl_string.rs (hunk 2 above)
Comment (bodhif4) [reply components/tidb_query_vec_expr/src/impl_string.rs:498]: OK, thanks. I want to try it.

**Comment by Fullstop000** [comment]
@TennyZhuang @breeswish Do we need to separate `lower` first?

**Pull request closed without merging**